File: mars/cs/mars.js
(function() {
  var o, r;
  o = 'NESW';
  r = function(p, i) {
    var k;
    if (i.length) {
      k = i.shift();
      if (k === 'M') {
        if (p[2] % 4 < 2) {
          p[p[2] % 2] += 1;
        } else {
          p[p[2] % 2] -= 1;
        }
      } else if (k === 'R') {
        p[2]--;
      } else {
        p[2]++;
      }
      return r(p, i);
    }
      
    return [p[1], p[0], o.charAt(p[2] % 4)];
  };
  console.log(r([0, 0, 1], 'MRMLM'.split('')));
}).call(this);

File: mars/py/mars.py
m='NESW'
v=raw_input
v()
def r(p,i):
  if i:x=i[0];p[[p[2]%2,2][x!='M']]+=[1,-1][x>'M']if x!='M'else p[2]%4<2 or-1
  return r(p,i[1:])if i else(p[1],p[0],m[p[2]%4])
try:
  while 1:x,y,z=v()[::2];print"%d %d %s"%r(map(int,[y,x,m.find(z)]),v())
except:1
